File: lib/diagnostic-utils.ts
/**
 * @file diagnostic-utils.ts
 * @description Helper utilities for diagnostic execution formatting, status telemetry, and metric computation.
 * Role: Provides core diagnostic primitives for system health monitoring and telemetry.
 * Integration: Used by diagnostic-engine.ts and kernel-lifecycle modules.
 */

import { DiagnosticSummary, TelemetryMetadata } from './diagnostic-types';

/**
 * Returns ISO 8601 formatted UTC timestamp.
 */
export function formatTimestamp(): string {
  return new Date().toISOString();
}

/**
 * Computes summary metrics for diagnostic check results.
 */
export function summarizeDiagnosticResults(checks: Record<string, boolean>): DiagnosticSummary {
  const total = Object.keys(checks).length;
  const passed = Object.values(checks).filter(Boolean).length;
  const failed = total - passed;
  const isHealthy = total > 0 && failed === 0;
  const passRate = total > 0 ? Math.round((passed / total) * 10000) / 100 : 0.0;

  return {
    total,
    passed,
    failed,
    isHealthy,
    passRate,
  };
}

/**
 * Executes a diagnostic check and measures execution duration in milliseconds.
 */
export async function executeCheckWithTelemetry(
  checkFn: () => Promise<boolean>
): Promise<{ passed: boolean; durationMs: number }> {
  const startTime = performance.now();
  try {
    const passed = await checkFn();
    const durationMs = performance.now() - startTime;
    return { passed, durationMs: Math.round(durationMs * 1000) / 1000 };
  } catch (e) {
    const durationMs = performance.now() - startTime;
    return { passed: false, durationMs: Math.round(durationMs * 1000) / 1000 };
  }
}

/**
 * Generates standard telemetry metadata for diagnostic results.
 */
export function generateTelemetryMetadata(): TelemetryMetadata {
  return {
    timestamp: Date.now(),
    version: "1.0.0-DIAGNOSTIC-AWARE",
    environment: typeof process !== 'undefined' ? process.env.NODE_ENV : 'unknown'
  };
}

File: Concept/tessera-enterprise/Modules/Calculator/diagnostic_utils.py
import time
from typing import Dict, Any, Callable, Tuple

def format_timestamp() -> str:
    return time.strftime('%Y-%m-%dT%H:%M:%SZ', time.gmtime())

def execute_check_with_telemetry(check_fn: Callable[[], bool], check_type: str) -> Tuple[bool, float]:
    start = time.perf_counter()
    try:
        passed = bool(check_fn())
        duration = (time.perf_counter() - start) * 1000
        return passed, round(duration, 3)
    except Exception:
        return False, round((time.perf_counter() - start) * 1000, 3)

def summarize_results(checks: Dict[str, bool]) -> Dict[str, Any]:
    total = len(checks)
    passed = sum(1 for v in checks.values() if v)
    return {
        'total': total,
        'passed': passed,
        'failed': total - passed,
        'is_healthy': total > 0 and passed == total
    }